File: internal/content/testdata/bench/javascript/nc-min-window-substring.js
function min_window(s, t) {
    if (!t || !s) return "";
    const need = new Map();
    for (const ch of t) {
        need.set(ch, (need.get(ch) || 0) + 1);
    }
    const have = new Map();
    let formed = 0;
    const required = need.size;
    let left = 0;
    let bestLen = Infinity;
    let bestStart = 0;
    for (let right = 0; right < s.length; right++) {
        const ch = s[right];
        have.set(ch, (have.get(ch) || 0) + 1);
        if (need.has(ch) && have.get(ch) === need.get(ch)) {
            formed++;
        }
        while (formed === required) {
            if (right - left + 1 < bestLen) {
                bestLen = right - left + 1;
                bestStart = left;
            }
            const lch = s[left];
            have.set(lch, have.get(lch) - 1);
            if (need.has(lch) && have.get(lch) < need.get(lch)) {
                formed--;
            }
            left++;
        }
    }
    if (bestLen === Infinity) return "";
    return s.substring(bestStart, bestStart + bestLen);
}

File: internal/content/testdata/bench/java/nc-min-window-substring.java
import java.util.*;
class Solution {
    public String min_window(String s, String t) {
        if (t.isEmpty() || s.isEmpty()) return "";
        Map<Character, Integer> need = new HashMap<>();
        for (char c : t.toCharArray()) need.put(c, need.getOrDefault(c, 0) + 1);
        Map<Character, Integer> have = new HashMap<>();
        int formed = 0, required = need.size();
        int left = 0, bestLen = Integer.MAX_VALUE, bestStart = 0;
        for (int right = 0; right < s.length(); right++) {
            char ch = s.charAt(right);
            have.put(ch, have.getOrDefault(ch, 0) + 1);
            if (need.containsKey(ch) && have.get(ch).equals(need.get(ch))) formed++;
            while (formed == required) {
                if (right - left + 1 < bestLen) { bestLen = right - left + 1; bestStart = left; }
                char lc = s.charAt(left);
                have.put(lc, have.get(lc) - 1);
                if (need.containsKey(lc) && have.get(lc) < need.get(lc)) formed--;
                left++;
            }
        }
        return bestLen == Integer.MAX_VALUE ? "" : s.substring(bestStart, bestStart + bestLen);
    }
}
